add id token support to verify access token hashes, fixes #126

gotwarlost opened the pull request.

ericchiang commented on oidc.go:
```diff
@@ -31,6 +35,10 @@ const (
 
+// ErrNoAccessTokenHash is returned when the caller attempts to validate an
+// access token hash against an id token that doesn't carry a value for it.
+var ErrNoAccessTokenHash = errors.New("id token did not have an access token hash")
```
Make this private? Instead of users comparing a sentinel error, we should probably just have them do something like

```
if idToken.AccessTokenHash == "" {
     // handle case
}
```

ericchiang commented on oidc.go:
```diff
@@ -242,6 +250,13 @@ type IDToken struct {
 	// and it's the user's responsibility to ensure it contains a valid value.
 	Nonce string
 
+	// at_hash claim if set in the ID token.
```
Document that users should use `VerifyAccessTokenHash`

ericchiang commented on oidc.go:
```diff
@@ -267,13 +282,43 @@ func (i *IDToken) Claims(v interface{}) error {
+// don't match. It is the caller's responsibility to ensure that the optional
+// access token hash is present for the id token before calling this method.
+// See https://openid.net/specs/openid-connect-core-1_0.html#CodeIDToken
+func (i *IDToken) VerifyAccessTokenHash(accessToken string) error {
```
`VerifyAccessToken`? We're verifying the access token, not the hash.

ericchiang commented on oidc_test.go:
```diff
@@ -0,0 +1,108 @@
+				if err == nil {
+					return fmt.Errorf("expected error, got success")
+				}
+				if err.Error() != `VerifyAccessTokenHash: unsupported signing algorithm "none"` {
```
Don't depend on error strings from other libraries.

ericchiang commented on oidc.go:
```diff
@@ -267,13 +282,43 @@ func (i *IDToken) Claims(v interface{}) error {
+	case RS512, ES512, PS512:
+		h = sha512.New()
+	default:
+		return fmt.Errorf("VerifyAccessTokenHash: unsupported signing algorithm %q", i.sigAlgorithm)
```
nit: ` fmt.Errorf("oidc: unsupported signing algorithm %q", i.sigAlgorithm)`

ericchiang commented on oidc.go:
```diff
@@ -267,13 +282,43 @@ func (i *IDToken) Claims(v interface{}) error {
+	sum := h.Sum(nil)[:h.Size()/2]
+	actual := base64.RawURLEncoding.EncodeToString(sum)
+	if actual != i.AccessTokenHash {
+		return fmt.Errorf("invalid access token hash, want %q, got %q", i.AccessTokenHash, actual)
```
I don't think we need to print "want" and "got" Users don't get much from a giant hash value.

gotwarlost replied:
agreed but might be useful for debugging further?

ericchiang replied:
> agreed but might be useful for debugging further?

We'd need a more information than two hashes to debug further. SIG alg, access token, etc, and we shouldn't return the access token in the error message because it's sensitive.

ericchiang commented:
Mostly nits. Other than that works for me.

gotwarlost commented:
addressed your comments (and improved the verify test to not pollute `verificationTest` with random methods). Thanks for being so quick with the review!

gotwarlost commented:
oops, didn't use my github id for the commits. I need to rewrite them.

gotwarlost committed:
add id token support to verify access token hashes, fixes #126

gotwarlost commented:
Unfortunately my commit rewrite lost the deltas that could be mapped to your comments. Sorry for the trouble!

ericchiang reviewed (approved):
lgtm

gotwarlost commented:
wow, that was fast! Thanks a bunch!
